chiron/mcmc.py

```diff
@@ -112,103 +175,802 @@ def run(
+        report_frequency
+            Frequency of saving the simulation data.
+        update_stepsize
+            Whether to update the "stepsize" of the move. Stepsize is a generic term for the key move parameters.
+            For example, for a simple displacement move this would be the displacement_sigma.
+        update_stepsize_frequency
+            Frequency of updating the stepsize of the move.
+        method
+            Methodology to use for accepting or rejecting the proposed state.
+            Default is "metropolis".
         """
-        Apply a Monte Carlo move to the system.
+        super().__init__(
+            nr_of_moves,
+            reporter=reporter,
+            report_frequency=report_frequency,
+        )
+        self.method = method  # I think we should pass a class/function instead of a string, like space.
 
-        This method should be overridden by subclasses to define specific types of moves.
+        self.reset_statistics()
+        self.update_stepsize = update_stepsize
+        self.update_stepsize_frequency = update_stepsize_frequency
 
-        Raises
-        ------
-        NotImplementedError
-            If the method is not implemented in subclasses.
+    def update(
+        self,
+        sampler_state: SamplerState,
+        thermodynamic_state: ThermodynamicState,
+        nbr_list: Optional[PairsBase] = None,
+    ) -> Tuple[SamplerState, ThermodynamicState, Optional[PairsBase]]:
         """
+        Perform the defined move and update the state.
+
+        Parameters
+        ----------
+        sampler_state : SamplerState
+            The initial state of the simulation, including positions.
+        thermodynamic_state : ThermodynamicState
+            The thermodynamic state of the system, including temperature and potential.
+        nbr_list : PairBase, optional
+            Neighbor list for the system.
 
-        raise NotImplementedError("apply_move() must be implemented in subclasses")
+        Returns
+        -------
+        sampler_state : SamplerState
+            The updated sampler state.
+        thermodynamic_state : ThermodynamicState
+            The updated thermodynamic state.
+        nbr_list: PairsBase
+            The updated neighbor/pair list. If a nbr_list is not set, this will be None.
+        """
+        calculate_current_potential = True
+
+        for i in range(self.nr_of_moves):
+            sampler_state, thermodynamic_state, nbr_list = self._step(
+                sampler_state,
+                thermodynamic_state,
+                nbr_list,
+                calculate_current_potential=calculate_current_potential,
+            )
+            # after the first step, we don't need to recalculate the current potential, it will be stored
+            calculate_current_potential = False
 
-    def compute_acceptance_probability(
+            elapsed_step = i + self._move_iteration * self.nr_of_moves
```

Comment:
I moved the comment up.  Basically it is the total number of steps taken for a given move (this has "memory" and will continue to increase each time we call, rather than being reset).  
As I said in the comment, if we say we want to update the maximum displacement_sigma, every 1000 steps, but only run the MC displacement move for 100, we’d either never update if we have `if step > 0 and step%update == 0:` or updated more frequently if we didn’t have condition that `step > 0` and just looked at the modulus.

Reply:
I think this is a really bad idea, since `self.nr_of_moves` can be changed on the fly.
If we need a step accumulator for statistics or auto-tuning, let's actually use a step accumulator like `self._number_of_attempts_made` and simply increment it with each move attempt.